### ISSUE TITLE
Uncaptured vars in laws

### ISSUE THREAD

**Issue by durban** [typelevel/cats-effect#231]
[`SyncLaws#repeatedSyncEvaluationNotMemoized`](https://github.com/typelevel/cats-effect/blob/master/laws/shared/src/main/scala/cats/effect/laws/SyncLaws.scala#L78) creates a `var`, which is then modified by the `F[_]` which is returned. Creating the var is not inside a `delay` or `suspend`. This makes the returned `F[_]` runnable only once (or rather, the second run will reuse the already modified `var`) which results in correct code seeming to violate the law.

Other laws which seem suspect (although I didn't tested them explicitly):

- `SyncLaws`:
  - `bracketReleaseCalledForError`
  - `unsequencedDelayIsNoop` (although here it probably doesn't matter)
  - `bindSuspendsEvaluation`
  - `mapSuspendsEvaluation`
- `AsyncLaws`:
  - `repeatedAsyncEvaluationNotMemoized`
  - `repeatedCallbackIgnored`

**Comment by alexandru** [comment]
@durban this is true, we initially used vars. However recently I started working with `Pledge`, which was now replaced by `Deferred`. So I think most vars can be replaced either by `Deferred` or by `MVar`.

But somebody has to submit a PR.
Will you try your hand at it? You donțt have to fix all of them, just what you can.

Also make sure you don't introduce other restrictions.

**Comment by alexandru** [comment]
Also, in the case of `repeatedSyncEvaluationNotMemoized`, I think we can just `delay` that. So that's an easy fix.

**Comment by johnynek** [comment]
I think we need to be a little careful when we change these. Do we have tests using incorrect implementations that they do fail the tests? It would be good to have that test using a intentionally broken versions to make sure the tests don’t accidentally become no-ops.

**Comment by alexandru** [comment]
That's a very good point. Also, what happened before was that some laws where introducing non-obvious restrictions, by mistake.

So now whenever I'm the one doing the release, I publish a snapshot version first and then also check the laws against Monix — e.g. `Iterant` also implementing `Async` has been extremely valuable in protecting against overreaching laws.

Looking at `repeatedSyncEvaluationNotMemoized`, I think that it's safe to suspend that though.
So there are cases that can be easily solved.

But it's not a burning issue.

**Comment by djspiewak** [comment]
IMO, since we now have an entirely new set of laws, none of which contain uncaptured vars, we can probably safely close this.

**Comment by alexandru** [comment]
You're cleaning up old issues I see 🤗
